kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala

```diff
@@ -102,6 +104,20 @@ class SparkProcessBuilder(
     }
   }
 
+  override protected def engineScalaBinaryVersion: String =
+    StringUtils.defaultIfBlank(System.getenv("SPARK_SCALA_VERSION"), sparkCoreScalaVersion)
+
+  private def sparkCoreScalaVersion: String = {
+    val sparkCoreJarPattern = Pattern.compile("^spark-core_.*\\.jar$")
+    Paths.get(sparkHome, "jars").toFile
+      .list((_, fileName) => sparkCoreJarPattern.matcher(fileName).matches())
+      .map { p => p.substring(p.indexOf("_") + 1, p.lastIndexOf("-")) }
+      .headOption match {
+      case Some(scalaVersion) => scalaVersion
+      case None => scalaRuntimeSemanticVersion.toString
```

Comment:
unreachable code?

Reply:
It's for a falling back if the scala version is unparsable from the spark core jar.

Reply:
Or let me skip it.